### DISCUSSION TITLE
Back from Kotlin to Scala.

### DISCUSSION THREAD

**Discussion by wiltonlazary** [rssh/dotty-cps-async#37] [General]
Hi, congratulations for this super important project.

Back-end programming is absolutely about non-block IO and Kotlin nailed it with co-routines that can carry context with then and can be managed by a specific implementation if necessary.

I think that async/await is a construction  on more generalized core concepts like CPS.

Async and non-block are different concepts, with co-routines/CPS I can make a lib to free the developer of the burden of think on concepts of computation all the time and focus on problem solving and when a parallel or asynchronous action makes sense, then use async/await appropriately.

I made one some time ago: https://github.com/wiltonlazary/cps_ppx

I'm looking forward to see Scala supporting CPS as a core concept to relief the burden of call an async function when on the reality the programmer only want a non-block action.

Thanks.

**Comment by rssh** [comment]
Thanks for the message and sorry for the late reply. [Usually, if something requires thinking, then it's delayed near forever ;).   ] 
About non-blocking actions and 'a-la synchronous' flow control: 
A)  dotty-cps now with automatic coloring:  https://rssh.github.io/dotty-cps-async/Features.html#automatic-coloring .  It's working now with all common effect monads.
B).  In the future,  all JVM calls will be non-blocking  [Poject Loom, https://wiki.openjdk.java.net/display/loom/Main ]
C).  In scala.js, all operations are non-blocking by design.
D) Not sure what the situation will be with scala-native; I guess both dotty-cps-async and Koka-like continuation design, based on setjmp/longjmp, potentially possible, but now is too early to think about details.

**Reply by wiltonlazary** [reply]
Great news,
I used the project https://github.com/puniverse/quasar fibers, loom is taking forever but will be a nice movement.

Thanks for the insights.
